src/sdk/object_io.js

```diff
@@ -671,43 +724,57 @@ class ObjectIO {
 
 
 /**
+ * slice_buffers_in_range
+ * It also returns missing parts in addition to found parts(chunks)
  *
  * @param {nb.Chunk[]} chunks
  * @param {number} start
  * @param {number} end
- * @return {Buffer[]}
+ * @return {Array<{ start: number, end: number, data: Buffer | null }>}
  */
 function slice_buffers_in_range(chunks, start, end) {
     if (end <= start) {
         // empty read range
         return null;
     }
     if (!chunks || !chunks.length) {
-        dbg.error('no chunks for data', range_utils.human_range({ start, end }));
-        throw new Error('no chunks for data');
+        // null data means that the part is missing.
+        return [ { start, end, data: null } ];
     }
     let pos = start;
-    /** @type {Buffer[]} */
     const buffers = [];
     for (const chunk of chunks) {
         const part = chunk.parts[0];
         let part_range = range_utils.intersection(part.start, part.end, pos, end);
-        if (!part_range) continue;
+        if (!part_range) {
+            if (end <= part.start) {
+                // null data means that the part is missing.
+                buffers.push({ start: pos, end, data: null });
+                pos = end;
+                break;
+            }
+            continue;
+        }
+
+        if (pos < part_range.start) {
+            // null data means that the part is missing.
+            buffers.push({ start: pos, end: part_range.start, data: null });
+        }
+
         let buffer_start = part_range.start - part.start;
         let buffer_end = part_range.end - part.start;
         if (part.chunk_offset) {
             buffer_start += part.chunk_offset;
             buffer_end += part.chunk_offset;
         }
         pos = part_range.end;
-        buffers.push(chunk.data.slice(buffer_start, buffer_end));
+        buffers.push({ start: part_range.start, end: part_range.end, data: chunk.data.slice(buffer_start, buffer_end) });
+
+        if (pos >= end) break;
     }
-    if (pos !== end) {
-        dbg.error('missing parts for data',
-            range_utils.human_range({ start, end }),
-            'pos', pos, chunks
-        );
-        throw new Error('missing parts for data');
+    const last_chunk_end = chunks[chunks.length - 1].parts[0].end;
+    if ((pos < end && pos > start) || start >= last_chunk_end) {
```

Comment:
Why is the condition pos != end or pos<end not enough?
This check against the last chunk end seems a bit artificial.
Which case is this covering?
```javascript
// --- start ------------- pos -------------- end ---
//       |                  |                  |
//       |[  buf  ][  buf  ]|<-- no more parts | 
//       |                  |                  |
//       |                  |[    missing     ]|
```

Reply:
pos!==end should work as well.

Reply:
@paulxuzhang Oh, I meant why isn't pos != end sufficient as the only condition.
if pos === start (opposite of pos > start) it means that we did not find any part that intersects with start -  why would we skip adding a missing range for this case?
The condition on last_chunk_end is not clear to me what purpose it serves.

Reply:
The condition (start >= last_chunk_end) is to cover the scenario that all chunks are on the left of the range.
```
        // ------------------------start ------------ end ---
        //                           |                 |
        //   ...  [  buf  ][  buf  ] |<-- last part    |
        //                           |                 |
        //                           |[    missing    ]|
```

Reply:
Also updated the comment at https://github.com/noobaa/noobaa-core/blob/86b728eb98d54092e55437f9a0fa00f7b40d01cf/src/sdk/object_io.js#L763-L778

Reply:
But I still can't figure out why is that condition needed - since pos is initialized to start and moves forward based on intersecting parts, how would this condition ever be used? Seems to me that it is redundant and that the condition pos<end is enough to decide if a missing range should be added for the remaining range.

Reply:
We initialize pos to start. If all the chunks are at the left of the read range,the for loop does not add missing part, pos stays as start.

Reply:
Also If we replaced
```
if ((pos !== end && pos > start) || start >= last_chunk_end)
```
with
```
if (pos !== end)
```
,

if the read range is on the left of all chunks, missing range (entire read range) would be added twice. One in the for loop, the other one in this if block.